src/Microsoft.AspNet.Server.Kestrel/Http/Connection.cs

```diff
@@ -35,78 +35,80 @@ public interface IConnectionControl
-        private static readonly Func<UvStreamHandle, int, object, Libuv.uv_buf_t> _allocCallback = AllocCallback;
+        private readonly Action<int, Exception> _readCallback;
+        private readonly Func<int, UvBuffer> _allocCallback;
+        private readonly UvTcpStreamHandle _socket;
+        private readonly Listener _listener;
 
-        private static Libuv.uv_buf_t AllocCallback(UvStreamHandle handle, int suggestedSize, object state)
-        {
-            return ((Connection)state).OnAlloc(handle, suggestedSize);
-        }
-
-        private static void ReadCallback(UvStreamHandle handle, int nread, Exception error, object state)
-        {
-            ((Connection)state).OnRead(handle, nread, error);
-        }
-
-        private readonly UvStreamHandle _socket;
+        private UvReadHandle _read;
         private Frame _frame;
         long _connectionId;
+        private bool _isInKeepAlive;
 
-        public Connection(ListenerContext context, UvStreamHandle socket) : base(context)
+        public Connection(Listener listener, UvTcpStreamHandle socket) : base(listener)
         {
+            _readCallback = OnRead;
+            _allocCallback = OnAlloc;
             _socket = socket;
             ConnectionControl = this;
-        }
+            _listener = listener;
 
-        public void Start()
-        {
             KestrelTrace.Log.ConnectionStart(_connectionId);
 
             SocketInput = new SocketInput(Memory);
             SocketOutput = new SocketOutput(Thread, _socket);
             _frame = new Frame(this);
-            _socket.ReadStart(_allocCallback, _readCallback, this);
+            _read = new UvReadHandle(_socket, _allocCallback, _readCallback);
+            listener.AddConnection(this);
         }
 
-        private Libuv.uv_buf_t OnAlloc(UvStreamHandle handle, int suggestedSize)
+        private UvBuffer OnAlloc(int suggestedSize)
         {
-            return handle.Libuv.buf_init(
-                SocketInput.Pin(2048),
-                2048);
+            const int bufferSize = 2048;
+            return new UvBuffer(SocketInput.Pin(bufferSize), bufferSize);
         }
 
-        private void OnRead(UvStreamHandle handle, int status, Exception error)
+        private void OnRead(int status, Exception error)
         {
             SocketInput.Unpin(status);
 
             var normalRead = error == null && status > 0;
             var normalDone = status == 0 || status == -4077 || status == -4095;
-            var errorDone = !(normalDone || normalRead);
 
-            if (normalRead)
-            {
-                KestrelTrace.Log.ConnectionRead(_connectionId, status);
-            }
-            else if (normalDone || errorDone)
+            if (!normalRead)
             {
                 KestrelTrace.Log.ConnectionReadFin(_connectionId);
                 SocketInput.RemoteIntakeFin = true;
-                _socket.ReadStop();
+                if (status != -4095)
```

Comment:
Can you define a descriptive constant for this? The same goes for the ones above.